crypto/compactcert/weights.go

```diff
@@ -0,0 +1,175 @@
+	return (&big.Int{}).SetUint64(num)
+}
+
+// lnIntApproximation returns a uint64 approximation
+func lnIntApproximation(x uint64) (uint64, error) {
+	if x == 0 {
+		return 0, ErrIllegalInputForLnApprox
+	}
+	result := math.Log(float64(x))
+	precision := uint64(1 << precisionBits)
+	expandWithPrecision := result * float64(precision)
+	return uint64(math.Ceil(expandWithPrecision)), nil
+
+}
+
+// verifyWeights makes sure that the number of reveals in the cert is correct with respect
+// to the signedWeight and a provenWeight upper bound.
+// This function checks that the following inequality is satisfied
+//
+// numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+//
+// where signedWeight/(2^d) >=1 for some integer d>=0, p = P/(2^b) >= ln(provenWeight), t = T/(2^b) >= ln(2) >= (T-1)/(2^b)
+// for some integers P,T >= 0 and b=16
+// more details can be found on the Algorand's spec
+func verifyWeights(signedWeight uint64, lnProvenWeight uint64, numOfReveals uint64, strengthTarget uint64) error {
+	if numOfReveals > MaxReveals {
+		return ErrTooManyReveals
+	}
+
+	if signedWeight == 0 {
+		return ErrZeroSignedWeight
+	}
+
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	//  numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+	//        /\
+	//        ||
+	//		  \/
+	// numReveals * (x + w * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+	y, x, w := getSubExpressions(signedWeight)
+	lhs := &big.Int{}
+	lhs.Set(w).
+		Mul(lhs, y).
+		Add(x, lhs).
+		Mul(bigInt(numOfReveals), lhs)
+
+	revealsTimesP := &big.Int{}
+	revealsTimesP.Set(bigInt(numOfReveals)).Mul(revealsTimesP, bigInt(lnProvenWeight))
+
+	rhs := &big.Int{}
+	rhs.Set(bigInt(strengthTarget))
+	rhs.Mul(rhs, bigInt(ln2IntApproximation)).
+		Add(rhs, revealsTimesP).
+		Mul(rhs, y)
+
+	if lhs.Cmp(rhs) < 0 {
+		return ErrInsufficientSingedWeight
+	}
+
+	return nil
+}
+
+// numReveals computes the number of reveals necessary to achieve the desired
+// security parameters. we use value which satisfies the following equation:
+//
+// numReveals > = ((strengthTarget) * T * Y / (3 * 2^b * (signedWeight^2 - 2^2d) + (d * (T - 1) - P) * Y))
+func numReveals(signedWeight uint64, lnProvenWeight uint64, strengthTarget uint64) (uint64, error) {
+	// in order to make the code more readable and reusable we will define the following expressions:
+	// y = signedWeight^2 + 2^(d + 2) * signedWeight + 2^2d
+	// x = 3 * 2^b * (signedWeight^2 - 2^2d)
+	// w = d * (T - 1)
+	//
+	//  numReveals * (3 * 2^b * (signedWeight^2 - 2^2d) + d * (T-1) * Y) >= ((strengthTarget) * T + numReveals * P) * Y
+	//        /\
+	//        ||
+	//		  \/
+	// numReveals * (x + w * Y) >= ((strengthTarget) * T + numReveals * P) * Y
```

Comment:
Your suggestion looks great. 
I think still needs some more clarity in describing the formula implementation since the function calls implementing the formula is not straightforward. Please check my suggestions.